File: Backend/innkt.Messaging/src/config/kafkaConfig.js
const kafkaConfig = {
  bootstrapServers: process.env.KAFKA_BOOTSTRAP_SERVERS || 'localhost:9092',
  producer: {
    clientId: 'messaging-producer',
    acks: 'all',
    enableIdempotence: true,
    messageTimeoutMs: 30000,
    retryBackoffMs: 100,
    messageSendMaxRetries: 3,
    compressionType: 'snappy',
    batchSize: 16384,
    lingerMs: 5
  },
  consumer: {
    groupId: 'messaging-consumer-group',
    clientId: 'messaging-consumer',
    autoOffsetReset: 'earliest',
    enableAutoCommit: true,
    autoCommitIntervalMs: 5000,
    sessionTimeoutMs: 30000,
    heartbeatIntervalMs: 3000,
    maxPollIntervalMs: 300000,
    fetchMinBytes: 1,
    fetchWaitMaxMs: 100,
    stopOnError: false
  },
  topics: {
    // Notification topics
    notifications: 'notifications',
    userNotifications: 'user.notifications',
    
    // Social events
    socialEvents: 'social.events',
    followEvents: 'follow.events',
    postEvents: 'post.events',
    groupEvents: 'group.events',
    
    // Message events
    messageEvents: 'message.events',
    conversationEvents: 'conversation.events',
    
    // User events
    userEvents: 'user.events',
    userActivity: 'user.activity',
    
    // System events
    systemEvents: 'system.events',
    healthEvents: 'service.health'
  }
};

module.exports = kafkaConfig;

File: Backend/innkt.Notifications/Configuration/KafkaConfig.cs
using Confluent.Kafka;

namespace innkt.Notifications.Configuration;

/// <summary>
/// Kafka configuration for optimized notification delivery
/// </summary>
public static class KafkaConfig
{
    /// <summary>
    /// Kafka topics for different notification types
    /// </summary>
    public static class Topics
    {
        public const string UserNotifications = "user.notifications";
        public const string KidNotifications = "kid.notifications";
        public const string ParentNotifications = "parent.notifications";
        public const string RepostNotifications = "repost.notifications";
        public const string SafetyAlerts = "safety.alerts";
        public const string EducationalNotifications = "educational.notifications";
        public const string IndependenceDayNotifications = "independence.notifications";
        public const string GrokResponses = "grok.responses";
    }

    /// <summary>
    /// Get producer configuration with optimized settings
    /// </summary>
    public static ProducerConfig GetProducerConfig(string bootstrapServers)
    {
        return new ProducerConfig
        {
            BootstrapServers = bootstrapServers,
            Acks = Acks.All, // Ensure all replicas acknowledge
            EnableIdempotence = true, // Prevent duplicate messages
            MessageTimeoutMs = 30000,
            RequestTimeoutMs = 30000,
            CompressionType = CompressionType.Snappy, // Optimize for speed
            BatchSize = 16384, // 16KB batch size for efficiency
            LingerMs = 5, // Wait 5ms to batch messages
            SecurityProtocol = SecurityProtocol.Plaintext, // For local development
            StatisticsIntervalMs = 60000, // Statistics every minute
        };
    }

    /// <summary>
    /// Get consumer configuration with optimized settings
    /// </summary>
    public static ConsumerConfig GetConsumerConfig(string bootstrapServers, string groupId)
    {
        return new ConsumerConfig
        {
            BootstrapServers = bootstrapServers,
            GroupId = groupId,
            AutoOffsetReset = AutoOffsetReset.Earliest,
            EnableAutoCommit = false, // Manual commit for reliability
            SessionTimeoutMs = 30000,
            HeartbeatIntervalMs = 3000,
            MaxPollIntervalMs = 300000, // 5 minutes
            FetchMinBytes = 1,
            PartitionAssignmentStrategy = PartitionAssignmentStrategy.CooperativeSticky,
            SecurityProtocol = SecurityProtocol.Plaintext,
            StatisticsIntervalMs = 60000,
        };
    }
}
